docs: Add v7 migration guide (#218)
seek-oss/sku: docs/migration-guides/v7.0.0.md

# v7.0.0 Migration guide

> This release changes the way sku handles rendering html & asset generation for both server rendered and statically rendered apps.

Reviewer: ```suggestion
This release changes the way sku handles HTML rendering and asset generation for both server rendered and statically rendered apps.
```

# v7.0.0 Migration guide

This release changes the way sku handles rendering html & asset generation for both server rendered and statically rendered apps.

> Please read through [static-rendering](../static-rendering) before attempting migration.

Reviewer: ```suggestion
Please read through our newly added documentation on [static rendering](../static-rendering) before attempting migration.
```

## Static apps

> Build output folder structure has changed in some circumstances. After making the following changes, please check your `target` directory and update any testing and deployment strategies to suit. See [static-rendering](../static-rendering) for details on the new structure.

Reviewer: ```suggestion
Build output folder structure has changed in some circumstances. After making the following changes, please check your `target` directory and update any testing and deployment strategies to suit. See our documentation on [static rendering](../static-rendering) for details on the new structure.
```

> ### render

Reviewer: ```suggestion
### Rendering
```

> If your app renders multiple pages/routes then first open up your `sku.config.js`.

Reviewer: ```suggestion
If your app renders multiple pages/routes, first open up your `sku.config.js`.

```

> Using the new `routes` option, detail all your applications routes.

Reviewer: ```suggestion
Using the new `routes` option, detail all of your application's routes.
```

> Also you now **must** set a `publicPath`.

Reviewer: ```suggestion
Also, you now **must** set a `publicPath`.
```

```js
module.exports = {
  routes: [
    { name: 'homePage', route: '/' },
    { name: 'detailsPage', route: '/details' }
  ],
  publicPath: 'www.cdn.com/my-app' // not used local development
};
```

The old render entry API looks like the following.

```js
// old style render entry
import React from 'react';
import { renderToString } from 'react-dom/server';

export default ({ publicPath }) => {
  const renderHTML = route => `
    <!DOCTYPE html>
    <html>
      <head>
        <meta charset="UTF-8">
        <title>My Awesome Project</title>
        <meta name="viewport" content="width=device-width, initial-scale=1">
        <link rel="stylesheet" type="text/css" href="${publicPath}/style.css" />
      </head>
      <body>
        <div id="app">${renderToString(<App route={route} />)}</div>
        <script src="${publicPath}/main.js"></script>
      </body>
    </html>
  `;

  // return HTML for each route
  return {
    '/': renderHTML('/'),
    '/details': renderHTML('/details')
  };
};
```

> The render entry now requires two function. `renderApp`, which should render your React app. And `renderDocument` which should render the HTML document. They both will be called for each route you specify in `sku.config.js`.

Reviewer: ```suggestion
The render entry now requires two functions:
- `renderApp`, which should render your React app.
- `renderDocument` which should render the HTML document.

They will both be called for each route you specify in `sku.config.js`.
```

> Also `main.js` & `styles.css` are no longer created. Instead, render the provided `headTags` and `bodyTags` at the end of their respective sections.

Reviewer: ```suggestion
Also, `main.js` & `styles.css` are no longer created. Instead, render the provided `headTags` and `bodyTags` at the end of their respective sections of the document.
```

```js
// new style render entry
import React from 'react';
import { renderToString } from 'react-dom/server';
import App from './App';

export default {
  renderApp: ({ environment, site, route }) =>
    renderToString(<App route={route} />),

  renderDocument: ({ app, bodyTags, headTags }) => `
    <!DOCTYPE html>
    <html>
      <head>
        <meta charset="UTF-8">
        <title>My Awesome Project</title>
        <meta name="viewport" content="width=device-width, initial-scale=1">
        ${headTags}
      </head>
      <body>
        <div id="app">${app}</div>
        ${bodyTags}
      </body>
    </html>
  `
};
```

> ### locales

Reviewer: ```suggestion
### `locales`
```

> The `locales` config has been removed. In most cases apps will be able to rename this property to `sites`.

Reviewer: ```suggestion
The `locales` config has been removed. In most cases, apps will be able to rename this property to `sites`.
```

```js
module.exports = {
  locales: ['au', 'nz']
};
// becomes
module.exports = {
  sites: ['au', 'nz']
};
```

> Then in the render entry pass the site option through to the React app.

Reviewer: ```suggestion
Then, in the render entry, pass the `site` option through to the React app.
```

```js
import React from 'react';
import { renderToString } from 'react-dom/server';

import App from './App';

export default {
  renderApp: ({ site }) => {
    // renderApp is called for each site
    return renderToString(<App site={site} />);
  },

  renderDocument: ({ site, app, bodyTags, headTags }) => `
    <!DOCTYPE html>
    <html>
      <head>
        <meta charset="UTF-8">
        <title>My Awesome Project</title>
        <meta name="viewport" content="width=device-width, initial-scale=1">
        <script>
          // Ensure the site is visible on the client side
          window.APP_CONFIG = ${JSON.stringify({ site })};
        </script>
        ${headTags}
      </head>
      <body>
        <div id="app">${app}</div>
        ${bodyTags}
      </body>
    </html>
  `
};
```

Finally, in your client entry you can access the `site`.

```js
import React from 'react';
import { hydrate } from 'react-dom';
import App from './App';

hydrate(<App site={window.APP_CONFIG.site} />, document.getElementById('app'));
```

[React context](https://reactjs.org/docs/context.html) can ease the pain of forwarding config props through your app.

> ### env

Reviewer: ```suggestion
### `env`
```

> The `env` config has been deprecated as it conflicts with the way sku builds assets. The `env` option allowed apps to embed environment specific config into the assets, sku now encourages apps to create environment agnostic assets, moving the config management responsibility to the HTML file.

Reviewer: ```suggestion
The `env` config has been deprecated as it conflicts with the way sku builds assets. The `env` option allowed apps to embed environment specific config into the assets, whereas sku now encourages apps to create environment agnostic assets, moving the config management responsibility to the HTML file at run time.
```

  }
};

export default {
  renderApp: ({ environment }) => {
    // renderApp is called for each environment
    const { apiUrl } = config[environment];

    return renderToString(<App apiUrl={apiUrl} />);
  },

  renderDocument: ({ environment, app, bodyTags, headTags }) => `
    <!DOCTYPE html>
    <html>
      <head>
        <meta charset="UTF-8">
        <title>My Awesome Project</title>
        <meta name="viewport" content="width=device-width, initial-scale=1">
        <script>
          // Ensure all the app config is visible on the client side
          window.APP_CONFIG = ${JSON.stringify(config[environment])};
        </script>
        ${headTags}
      </head>
      <body>
        <div id="app">${app}</div>
        ${bodyTags}
      </body>
    </html>
  `
};
```

Finally, in your client entry you can access the app config.

```js
import React from 'react';
import { hydrate } from 'react-dom';
import App from './App';

const config = window.APP_CONFIG;

hydrate(<App apiUrl={config.apiUrl} />, document.getElementById('app'));
```

[React context](https://reactjs.org/docs/context.html) can ease the pain of forwarding config props through your app.

Breaking changes:

> - Build output structure has changed and could break current deployment strategies

Reviewer: ```suggestion
- Build output structure has changed and could break current deployment strategies.
```

> - Update SSR port config

Reviewer: Is this meant to be here?

Author: Woops

## Server rendered apps

> ### port

Reviewer: ```suggestion
### Port
```

Specifying `port` as an object is no longer valid. Please specify `port` and `serverPort` individually if you want to override the defaults.

```js
module.exports = {
  port: { client: 1234, backend: 5678 }
};
// becomes
module.exports = {
  port: 1234,
  serverPort: 5678
};
```

> ### server entry

Reviewer: ```suggestion
### Server entry
```
